[CBRD-26275] ha 실행 시 RK 제약 조건 확인

youngjun9072 opened the pull request:
<http://jira.cubrid.org/browse/CBRD-26275>[CBRD-26275](http://jira.cubrid.org/browse/CBRD-26275)

### Purpose

HA 실행 시 RK 제약조건 확인 로직(유틸리티) 개발 이슈입니다.
유틸리티 이름은 rkcheck 입니다.

### Implementation

1. 복제 키 제약 조건을 확인하기위한 유틸리티 추가 이슈이며, HA 실행 시에도 heart beat에 의해 호출됩니다.
1-1. 복제 옵션이 on 인 테이블은 PK 혹은 NOT NULL UNIQUE KEY를 가져야합니다
1-2. 외래 키 참조시 복제 테이블 옵션이 off인 경우는 참조할 수 없습니다.

2. 이 로직은 hb start시 수행되는 유틸리티 호출 로직을 제외하면, 유틸리티만의 동작 로딩은 아래 절차로 수행됩니다.
2-1. 파라미터 확인
2-2. 에러 로그 파일 오픈
2-3. 클래스 목록 조회
2-4. 클래스 순회
2-5. 클래스의 attribute 순회
2-6. 제약 조건 확인
2-7. 위반 발견 시 로깅 및 에러 처리

3. 제약사항 위반 발견 시 다음 3가지 방법의 로깅 및 출력을 수행합니다.
3-1. 터미널 출력
```
@ cubrid heartbeat start 
@ cubrid master start 
++ cubrid master start: success 
@ HA processes start 
@ cubrid server start: demodb

This may take a long time depending on the amount of recovery works to do.

CUBRID 11.5.0 (11.5.0.1945) (64 debug build)

++ cubrid server start: success 
@ rkcheck start 
- check ha constraints: failed. 
- Replication Key violations: 1 
- Foreign Key violations: 1 
++ rkcheck start: fail 
++ HA processes start: fail 
++ cubrid heartbeat start: fail
```

3-2. 유틸리티 로그
path : $CUBRID/log/{database_name}_rkcheck.err
```
Time: 12/04/25 03:12:57.456 - DEBUG *** file /home/youngjun/cubrid/src/transaction/boot_cl.c, line 1134
boot_restart_client: register client { type 7 db demodb user DBA password (null) program rkcheck login youngjun host youngjun1 pid 3720220 }

Time: 12/04/25 03:12:57.494 - ERROR *** file /home/youngjun/cubrid/src/executables/util_cs.c, line 2839 ERROR CODE = -1371, Tran = 1, EID = 2
HA replication constraint violation: 2 constraint(s) failed, so HA mode cannot be started. Refer to '/home/youngjun/CUBRID-11.5.0.1944-dc40730-Linux.x86_64-debug/log/demodb_rkcheck_20260115_1426.list' for more details.

```

3-3. 제약사항 위반 리스트
path : $CUBRID/log/{dabase_name}_rkcheck_{timestamp}.list
```
< Replication Key(PRIMARY KEY or NOT NULL UNIQUE KEY) Constraint Violations >
dba.tbl


< Foreign Key Constraint Violations >
dba.b_tbl1(fk_id) -> dba.a_tbl1(pk_a_tbl1_id)


< HOW TO FIX >
* Replication Key violations
  - Run in single mode and add a PRIMARY KEY or a NOT NULL UNIQUE constraint to the table.
  - Alternatively, set the table's replication option to OFF (data in the table will not be replicated).
* Foreign Key violations
  - Run in single mode and remove the foreign key, OR
  - Change the replication option of the referenced table to ON.
```
### 테스트

1. RK 위반 없음
* 터미널 출력
```
[youngjun@youngjun3 ~]$ cub_admin rkcheck demodb

Constraint checks completed for all replication-enabled tables.
```
* .err 파일 출력
```
관련 출력 없음
```

*.list 파일 출력
```
< Replication Key(PRIMARY KEY or NOT NULL UNIQUE KEY) Constraint Violations >
All replicated tables comply with the replication key constraints.


< Foreign Key Constraint Violations >
All replicated tables comply with the foreign key constraints.
```

2. RK 위반
* csql
```
CREATE TABLE tbl(
    a INT
);
```

* 터미널 출력
```
[youngjun@youngjun3 ~]$ cub_admin rkcheck demodb
 - check ha constraints: failed.
 - Replication Key violations: 1
 - Foreign Key violations: 0
```

* .err 파일 출력
```
Time: 01/21/26 01:38:11.673 - ERROR *** file /home/youngjun/cubrid/src/executables/util_cs.c, line 3002 ERROR CODE = -1371, Tran = 1, EID = 2
HA replication constraint violation: 1 constraint(s) failed, so HA mode cannot be started. Refer to '/home/youngjun/CUBRID-11.5.0.1946-33cc19b-Linux.x86_64-debug/log/demodb_rkcheck_20260121_0138.list' for more details.
```

* .list 파일 출력
```
< Replication Key(PRIMARY KEY or NOT NULL UNIQUE KEY) Constraint Violations >
dba.tbl


< Foreign Key Constraint Violations >
All replicated tables comply with the foreign key constraints.


< HOW TO FIX >
* Replication Key violations
  - Run in single mode and add a PRIMARY KEY or a NOT NULL UNIQUE constraint to the table.
  - Alternatively, set the table's replication option to OFF (data in the table will not be replicated).
* Foreign Key violations
  - Run in single mode and remove the foreign key, OR
  - Change the replication option of the referenced table to ON.
```

3. FK 위반
* csql
```
CREATE TABLE a_tbl1 (
    id INT NOT NULL DEFAULT 0 PRIMARY KEY,
    phone VARCHAR(10)
)REPLICATION OFF;

CREATE TABLE b_tbl1 (
    id INT NOT NULL,
    name VARCHAR (10) NOT NULL,
    CONSTRAINT pk_id PRIMARY KEY (id),
    CONSTRAINT fk_id FOREIGN KEY (id) REFERENCES a_tbl1 (id)
    ON DELETE CASCADE ON UPDATE RESTRICT
);
```

* 터미널 출력
```
[youngjun@youngjun3 ~]$ cub_admin rkcheck demodb
 - check ha constraints: failed.
 - Replication Key violations: 0
 - Foreign Key violations: 1
```
* .err 파일 출력
```
Time: 01/21/26 01:46:21.683 - ERROR *** file /home/youngjun/cubrid/src/executables/util_cs.c, line 3002 ERROR CODE = -1371, Tran = 1, EID = 2
HA replication constraint violation: 1 constraint(s) failed, so HA mode cannot be started. Refer to '/home/youngjun/CUBRID-11.5.0.1946-33cc19b-Linux.x86_64-debug/log/demodb_rkcheck_20260121_0146.list' for more details.
```
* .list 파일 출력
```
< Replication Key(PRIMARY KEY or NOT NULL UNIQUE KEY) Constraint Violations >
All replicated tables comply with the replication key constraints.


< Foreign Key Constraint Violations >
dba.b_tbl1(fk_id) -> dba.a_tbl1(pk_a_tbl1_id)


< HOW TO FIX >
* Replication Key violations
  - Run in single mode and add a PRIMARY KEY or a NOT NULL UNIQUE constraint to the table.
  - Alternatively, set the table's replication option to OFF (data in the table will not be replicated).
* Foreign Key violations
  - Run in single mode and remove the foreign key, OR
  - Change the replication option of the referenced table to ON.
```

4. 두 제약 모두 위반
* csql 
```
CREATE TABLE tbl(
    a INT
);

CREATE TABLE a_tbl1 (
    id INT NOT NULL DEFAULT 0 PRIMARY KEY,
    phone VARCHAR(10)
)REPLICATION OFF;

CREATE TABLE b_tbl1 (
    id INT NOT NULL,
    name VARCHAR (10) NOT NULL,
    CONSTRAINT pk_id PRIMARY KEY (id),
    CONSTRAINT fk_id FOREIGN KEY (id) REFERENCES a_tbl1 (id)
    ON DELETE CASCADE ON UPDATE RESTRICT
);
```

* 터미널 출력
```
[youngjun@youngjun3 ~]$ cub_admin rkcheck demodb
 - check ha constraints: failed.
 - Replication Key violations: 1
 - Foreign Key violations: 1
```

* .err 파일 출력
```
Time: 01/21/26 01:51:00.966 - ERROR *** file /home/youngjun/cubrid/src/executables/util_cs.c, line 3002 ERROR CODE = -1371, Tran = 1, EID = 2
HA replication constraint violation: 2 constraint(s) failed, so HA mode cannot be started. Refer to '/home/youngjun/CUBRID-11.5.0.1946-33cc19b-Linux.x86_64-debug/log/demodb_rkcheck_20260121_0151.list' for more details.
```

* .list 파일 출력
```
< Replication Key(PRIMARY KEY or NOT NULL UNIQUE KEY) Constraint Violations >
dba.tbl


< Foreign Key Constraint Violations >
dba.b_tbl1(fk_id) -> dba.a_tbl1(pk_a_tbl1_id)


< HOW TO FIX >
* Replication Key violations
  - Run in single mode and add a PRIMARY KEY or a NOT NULL UNIQUE constraint to the table.
  - Alternatively, set the table's replication option to OFF (data in the table will not be replicated).
* Foreign Key violations
  - Run in single mode and remove the foreign key, OR
  - Change the replication option of the referenced table to ON.
```

5. -r 옵션
* csql 
```
테스트 4와 동일
```

* 터미널 출력
```
[youngjun@youngjun3 ~]$ cub_admin rkcheck -r demodb
 - check ha constraints: failed.
 - Replication Key violations: 1
```

* .err 파일 출력
```
Time: 01/21/26 01:52:18.684 - ERROR *** file /home/youngjun/cubrid/src/executables/util_cs.c, line 3002 ERROR CODE = -1371, Tran = 1, EID = 2
HA replication constraint violation: 2 constraint(s) failed, so HA mode cannot be started. Refer to '/home/youngjun/CUBRID-11.5.0.1946-33cc19b-Linux.x86_64-debug/log/demodb_rkcheck_20260121_0152.list' for more details.
```

* .list 파일 출력
```
< Replication Key(PRIMARY KEY or NOT NULL UNIQUE KEY) Constraint Violations >
dba.tbl


< HOW TO FIX >
* Replication Key violations
  - Run in single mode and add a PRIMARY KEY or a NOT NULL UNIQUE constraint to the table.
  - Alternatively, set the table's replication option to OFF (data in the table will not be replicated).
```

6. -f 옵션
* csql 
```
테스트 4와 동일
```

* 터미널 출력
```
[youngjun@youngjun3 ~]$ cub_admin rkcheck -f demodb
 - check ha constraints: failed.
 - Foreign Key violations: 1
```

* .err 파일 출력
```
Time: 01/21/26 02:01:01.904 - ERROR *** file /home/youngjun/cubrid/src/executables/util_cs.c, line 3002 ERROR CODE = -1371, Tran = 1, EID = 2
HA replication constraint violation: 2 constraint(s) failed, so HA mode cannot be started. Refer to '/home/youngjun/CUBRID-11.5.0.1946-33cc19b-Linux.x86_64-debug/log/demodb_rkcheck_20260121_0201.list' for more details.
```

* .list 파일 출력
```
< Foreign Key Constraint Violations >
dba.b_tbl1(fk_id) -> dba.a_tbl1(pk_a_tbl1_id)


< HOW TO FIX >
* Foreign Key violations
  - Run in single mode and remove the foreign key, OR
  - Change the replication option of the referenced table to ON.
```

7. rk 제약사항만 위반이지만 -f 옵션 입력 시
* 터미널 출력
```
[youngjun@youngjun3 ~]$ cub_admin rkcheck -f demodb

Constraint checks completed for all replication-enabled tables.
```

* .err
```
관련 에러 없음
```

* .list
```
< Foreign Key Constraint Violations >
All replicated tables comply with the foreign key constraints.
```

8. fk제약사항만 위반이지만 -r 옵션 입력 시
* 터미널 출력
```
[youngjun@youngjun3 ~]$ cub_admin rkcheck -r demodb

Constraint checks completed for all replication-enabled tables.
```

* .err
```
관련 에러 없음
```

* .list
```
< Replication Key(PRIMARY KEY or NOT NULL UNIQUE KEY) Constraint Violations >
All replicated tables comply with the replication key constraints.
```

9. 잘못된 데이터 베이스
* 터미널 출력
```
[youngjun@youngjun3 ~]$ cub_admin rkcheck unknown
rkcheck: Failed to connect to database server, 'unknown', on the following host(s): localhost
```

* .err 
```
Time: 01/21/26 02:59:53.766 - ERROR *** file /home/youngjun/cubrid/src/transaction/boot_cl.c, line 1739 ERROR CODE = -677, Tran = -1, EID = 2
Failed to connect to database server, 'unkown', on the following host(s): localhost
```

* .list
```
파일 생성 없음
```

10. 잘못된 옵션(usage)
* 터미널
```
[youngjun@youngjun3 ~]$ cub_admin rkcheck -t demodb
rkcheck: invalid option -- 't'
rkcheck: Checks HA replication key constraints.
usage: cub_admin rkcheck [OPTION] database-name

valid option:
    -r, --rk-results           print RK(Replication Key) constraint violation list
    -f, --fk-results           print FK(Foreign Key) constraint violation list
```

* .err
```
관련 출력 없음
```

* .list
```
파일 생성 없음
```

### Remarks

youngjun9072 committed:
Introduce the rkcheck admin utility to check replication key constraints before starting HA processes.

youngjun9072 committed:
refoctoring

youngjun9072 committed:
commit

youngjun9072 committed:
fix typo

youngjun9072 committed:
Separate the list of constraint-violating tables into a dedicated file

youngjun9072 committed:
change section title

hornetmj commented on msg/ko_KR.utf8/cubrid.msg:
```diff
@@ -1463,8 +1463,9 @@ $ LOADDB
 1368 HA 모드에서는 테이블 생성 시 복제 후보 키(기본 키 또는 NOT NULL UNIQUE 제약)가 필요합니다. 없을 경우 REPLICATION 옵션을 OFF로 설정하십시오.
 1369 HA 모드에서는 클래스의 복제 옵션을 변경할 수 없습니다.
 1370 HA 복제 제약 위반: 외래 키를 가진 테이블은 참조하는 테이블과 동일한 복제 옵션을 가져야 합니다.
+1371 HA 복제 제약 위반: HA 복제 제약 위반: %1$d개의 제약이 실패하여 HA 모드를 시작할 수 없습니다. 자세한 내용은 '%2$s'을(를) 참고하세요.
```
```suggestion
1371 HA 복제 제약 위반: %1$d개의 복제 제약 위반이 발생하여 HA 모드를 시작할 수 없습니다. 자세한 내용은 '%2$s'을(를) 참고하세요.
```

youngjun9072 replied:
@hornetmj 반영했습니다.

hornetmj commented on src/executables/utility.h:
```diff
@@ -931,6 +933,7 @@ typedef struct _ha_config
 #define PRINT_CMD_GETID         "getid"
 #define PRINT_CMD_TEST          "test"
 #define PRINT_CMD_REPLICATION	"replication"
+#define PRINT_CMD_RKCHECK      "rkcheck"
```
```suggestion
#define PRINT_CMD_RKCHECK       "check_repl_cons"
```

youngjun9072 replied:
@hornetmj 반영했습니다.

hornetmj commented on src/executables/utility.h:
```diff
@@ -1020,6 +1023,7 @@ typedef struct _ha_config
 #define UTIL_OPTION_TDE			        "tde"
 #define UTIL_OPTION_FLASHBACK                   "flashback"
 #define UTIL_OPTION_MEMMON                      "memmon"
+#define UTIL_OPTION_RKCHECK                    "rkcheck"
```
```suggestion
#define UTIL_OPTION_RKCHECK                     "rkcheck"
```

youngjun9072 replied:
@hornetmj 반영했습니다.

hornetmj commented on src/executables/utility.h:
```diff
@@ -805,7 +805,8 @@ typedef enum
   FLASHBACK,
   MEMMON,
-  LOGFILEDUMP,
+  RKCHECK,
```
```suggestion
  CHECK_REPL_CONS,
```

youngjun9072 replied:
@hornetmj 반영했습니다.

hornetmj commented on src/executables/utility.h:
```diff
@@ -1756,6 +1760,12 @@ typedef struct _ha_config
+#define RKCHECK_USER_S             'u'
+#define RKCHECK_USER_L             "user"
+#define RKCHECK_PASSWORD_S         'p'
+#define RKCHECK_PASSWORD_L         "password"
```
```suggestion
#define RKCHECK_USER_S              'u'
#define RKCHECK_USER_L              "user"
#define RKCHECK_PASSWORD_S          'p'
#define RKCHECK_PASSWORD_L          "password"
```

youngjun9072 replied:
@hornetmj 삭제된 코드입니다.

hornetmj commented on src/executables/util_service.c:
```diff
@@ -222,6 +223,7 @@ static UTIL_SERVICE_OPTION_MAP_T us_Service_map[] = {
 #define COMMAND_TYPE_TEST       "test"
 #define COMMAND_TYPE_REPLICATION	"replication"
 #define COMMAND_TYPE_REPLICATION_SHORT	"repl"
+#define COMMAND_TYPE_RKCHECK   "rkcheck"
```
```suggestion
#define COMMAND_TYPE_RKCHECK    "rkcheck"
```

youngjun9072 replied:
@hornetmj 반영했습니다.

hornetmj commented on src/executables/util_service.c:
```diff
@@ -93,7 +93,8 @@ typedef enum
   TEST,
-  REPLICATION
+  REPLICATION,
+  SC_RKCHECK
```
us_Command_map[]의 RKCHECK를 SC_RKCHECK로 변경 또는 SC_RKCHECK를 RKCHECK로 변경; SC_ 로 특별한 구분을 하는 것 같지 않아 다음과 같이 제안은 드립니다.
```suggestion
  RKCHECK
```

youngjun9072 replied:
@hornetmj 반영했습니다.

hornetmj replied:
NIT:

요청건이 반영되지 않았습니다. 수정 여부는 판단하셔서 결정해 주시면 되겠습니다.

hornetmj commented on src/executables/util_service.c:
```diff
@@ -3205,6 +3211,35 @@ ha_argv_to_args (char *args, int size, const char **argv, HB_PROC_TYPE type)
+
+  for (i = 0; dbs[i] != NULL; i++)
+    {
+      db_name = dbs[i];
```
db_name이 지정된 경우 ha_conf에 설정된 dbs 중 db_name과 일치하는 부분만 실행시키면 될 것 같습니다. us_hb_copylogdb_start(), us_hb_applylogdb_start() 참고하시면 되겠습니다. (db_name != NULL && strcmp (dbs[i], db_name) != 0)

youngjun9072 replied:
@hornetmj 반영했습니다.

hornetmj commented on src/executables/util_service.c:
```diff
@@ -3205,6 +3211,35 @@ ha_argv_to_args (char *args, int size, const char **argv, HB_PROC_TYPE type)
+  for (i = 0; dbs[i] != NULL; i++)
+    {
+      db_name = dbs[i];
+      const char *lw_argv[] = { UTIL_ADMIN_NAME, UTIL_RKCHECK,
```
NIT:
lw_는 log writer(== copylogdb)의 prefix 입니다. 
```suggestion
      const char *rkcheck_args[] = { UTIL_ADMIN_NAME, UTIL_RKCHECK,
```

youngjun9072 replied:
@hornetmj 반영했습니다.

hornetmj commented on src/executables/util_service.c:
```diff
@@ -5131,6 +5175,7 @@ process_heartbeat (int command_type, int argc, const char **argv)
       break;
     case SC_COPYLOGDB:
     case SC_APPLYLOGDB:
+    case SC_RKCHECK:
```
cubrid hb(heartbeat) rkcheck 방식의 명령어가 수행될 수 있도록 허용하고 있습니다. 구동 시점에 내부적으로 호출되는 경우를 제외하고는 외부로 별도의 제어권을 넘겨줄 필요는 없을 것 같습니다.

youngjun9072 replied:
@hornetmj 반영했습니다.

hornetmj commented on src/executables/util_cs.c:
```diff
@@ -2727,6 +2746,104 @@ changemode (UTIL_FUNCTION_ARG * arg)
+      user = (char *) "DBA";
+    }
+
+  error = db_restart_ex (arg->command_name, database_name, user, password, NULL, DB_CLIENT_TYPE_ADMIN_UTILITY);
```
다음 코드를 참고하셔서 패스워드 없이 처리하실 수 있도록 하시면 되겠습니다.
```
  AU_DISABLE_PASSWORDS ();      /* disable authorization for this operation */
  db_set_client_type (DB_CLIENT_TYPE_ADMIN_UTILITY);
  db_login ("DBA", NULL);

  if (db_restart (arg->command_name, TRUE, database_name))
```

youngjun9072 replied:
@hornetmj 반영했습니다.

hornetmj commented on src/executables/util_cs.c:
```diff
@@ -2727,6 +2746,104 @@ changemode (UTIL_FUNCTION_ARG * arg)
+  /* error message log file */
+  snprintf (er_msg_file, sizeof (er_msg_file) - 1, "%s_%s.err", database_name, arg->command_name);
+  er_init (er_msg_file, ER_NEVER_EXIT);
+  er_set_print_property (ER_PRINT_TO_CONSOLE);
```
rkcheck 실행 중 발생하는 모든 에러가 콘솔로 출력될 것 같은데, 어떤 의도가 있으신 건지 추가 설명 요청드립니다.

youngjun9072 replied:
@hornetmj 처음 설계단계부터 터미널과 util의 .err로그 파일에 동일하게 남기는 것으로 설계되어 콘솔과 파일 둘다 남길수 있도록 er_set_print_property() 설정을 했습니다.

hornetmj replied:
er_set_print_property 사용이 제거된 것으로 보여 resolve 처리합니다.

hornetmj commented on src/executables/utility.h:
```diff
@@ -1756,6 +1760,12 @@ typedef struct _ha_config
+#define RKCHECK_USER_S             'u'
+#define RKCHECK_USER_L             "user"
+#define RKCHECK_PASSWORD_S         'p'
+#define RKCHECK_PASSWORD_L         "password"
```
사용되지 않는 옵션으로 보입니다.
```suggestion
```

youngjun9072 replied:
@hornetmj 제거했습니다.

hornetmj commented on src/executables/util_cs.c:
```diff
@@ -2727,6 +2746,104 @@ changemode (UTIL_FUNCTION_ARG * arg)
+  if (violation_count > 0)
+    {
+      er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_HA_REPLICATION_CONSTRAINT_VIOLATION, 2, violation_count,
+	      er_msg_file);
```
PR에 설명하신 대로 구현은 되어 있지만, '.list 파일'을 참조하라고 출력하는 것을 의도하신게 아닌가 생각됩니다.
```suggestion
	      violation_list_file);
```

youngjun9072 replied:
@hornetmj 잘못 작성된 변수명이라 violation_list_file로 변경했습니다.

hornetmj commented on src/executables/util_cs.c:
```diff
@@ -81,6 +82,24 @@
+  while (0)
+
+
+#define RK_CONSTRAINT_VIOLATIONS_SECTION_TITLE  "Replication Key(PRIMARY KEY or NOT NULL UNIQUE KEY) Constraint Violations"
```
솔루션 명시하는게 명확하지 않을까 생각합니다. 'ex) 다음 테이블(클래스)들에 복제키를 추가하거나 복제 대상에서 제외' 등)

구두 논의하겠습니다.

hornetmj commented on src/query/execute_schema.h:
```diff
@@ -107,5 +107,6 @@ extern int do_recreate_filter_index_constr (PARSER_CONTEXT * parser, SM_PREDICAT
 			     int *assigns_count, CLIENT_UPDATE_CLASS_INFO ** cls_data, int *cls_count,
 			     DB_VALUE ** values, int *values_cnt, bool has_delete);
-
+extern int check_ha_repl_constraint (DB_OBJECT * class_obj);
```
외부에서 호출되지 않습니다.

youngjun9072 replied:
@hornetmj  반영했습니다. execute_schema.c로 이동했고 static 함수로 변경했습니다.

youngjun9072 committed:
Apply review feedback for rkcheck and HA replication constraints

- Allow DBA login without password
- Rename violation list output from .err (er_msg_file) to .list (violation_list_file)
- Rename constants and remove unused ones
- Add spacing for better alignment and readability
- Check replication key constraints only for the db_name specified in config
  (check all configured databases if db_name is not provided)
- Rename variables for clarity
- Disallow executing rkcheck via "cubrid hb rkcheck"
- Change PRINT_CMD_RKCHECK value to "check_repl_cons"
- Change error message

youngjun9072 committed:
rkcheck: apply review feedback

- Add 'r' and 'f' options
- Revise error reporting: brief output to terminal, detailed output to file
- Change violation list file output format
- Add timestamp to .list file name

H2SU commented on src/query/execute_schema.c:
```diff
@@ -392,8 +392,8 @@ static int do_recreate_saved_indexes (MOP classmop, SM_CONSTRAINT_INFO * index_s
 
-static int check_ha_repl_constraint (DB_OBJECT * class_obj);
 static bool check_ha_repl_fk_ref_all_replicated (DB_OBJECT * class_obj);
+static int check_ha_repl_constraint (DB_OBJECT * class_obj);
```
위치만 위에서 아래로 변경된 코드로 보입니다.

youngjun9072 replied:
@H2SU 변경되었습니다.

H2SU commented on src/executables/util_cs.c:
```diff
@@ -2727,6 +2767,211 @@ changemode (UTIL_FUNCTION_ARG * arg)
+	   basename (arg->argv0));
+  util_log_write_errid (MSGCAT_UTIL_GENERIC_INVALID_ARGUMENT);
+
+  return EXIT_FAILURE;
```
```suggestion
  return err;
```

youngjun9072 replied:
@H2SU 변경하였습니다.

hgryoo commented on msg/en_US.utf8/utils.msg:
```diff
@@ -1336,3 +1336,12 @@ usage: %1$s memmon [OPTION] database-name\n\
+\n\
+valid option:\n\
+    -r, --rk-result           print rk result\n\
+    -f, --fk-results          print fk result\n
```
nit: newline
```suggestion
    -f, --fk-results          print fk result\n

```

hornetmj replied:
--rk-result와 --fk-results의 단수 또는 복수로 통일

youngjun9072 replied:
@hgryoo @hornetmj 반영하였습니다. 복수로 변경하였고, 더불어 설명이 너무 간단한것같아 조금 더 추가했습니다.

hgryoo commented on msg/ko_KR.utf8/utils.msg:
```diff
@@ -1322,3 +1322,12 @@ memmon: 데이터베이스 서버의 메모리 사용량 정보를 제공합니
+\n\
+valid option:\n\
+    -r, --rk-result           rk 제약사항 확인 결과를 출력합니다\n\
+    -f, --fk-result           fk 제약사항 확인 결과를 출력합니다\n
```
nit: newline

```suggestion
    -f, --fk-result           fk 제약사항 확인 결과를 출력합니다\n

```

InChiJun commented on src/executables/util_cs.c:
```diff
@@ -2727,6 +2767,211 @@ changemode (UTIL_FUNCTION_ARG * arg)
 }
 
+static char *
+generate_violation_list_file_name (char *out, char *database_name, const char *util_name)
```
이 함수 선언부가 없는 것 같은데 의도하신걸까요?

InChiJun replied:
아래의 새로 추가된 함수들도 선언부가 따로 없습니다.
open_violation_list_file()
get_print_flags()
check_rk_constraint()
check_fk_constraint()

youngjun9072 replied:
@InChiJun 선언부 추가하였습니다.

InChiJun commented on src/executables/util_cs.c:
```diff
@@ -2727,6 +2767,211 @@ changemode (UTIL_FUNCTION_ARG * arg)
+	{
+	  fprintf (fp, "%s\n", sm_get_ch_name (classop));
+	}
+      return 1;
+    }
+
+  return 0;
```
retrun값 0과 1이 의미하는 바가 무엇인지 주석으로 표기하면 좋겠습니다.

youngjun9072 replied:
@InChiJun 주석 추가하였습니다.

InChiJun commented on src/executables/util_cs.c:
```diff
@@ -2727,6 +2767,211 @@ changemode (UTIL_FUNCTION_ARG * arg)
+}
+
+int
+rkcheck (UTIL_FUNCTION_ARG * arg)
```
함수명을 replication_key_check()와 같이 정확한 명칭을 지정하거나, 함수 주석을 추가하는게 함수 의도를 파악하기 쉬울 것 같습니다.

youngjun9072 replied:
@InChiJun 다른 유틸리티의 메인 함수처럼 유틸리티 명을 따랐습니다. 주석 추가했습니다.

hornetmj commented on msg/en_US.utf8/utils.msg:
```diff
@@ -1336,3 +1336,12 @@ usage: %1$s memmon [OPTION] database-name\n\
     -o, --output-file=FILE    output results to FILE\n\
     --disable-force           disable memory monitor of server forcibly\n
+
+$set 60 MSGCAT_UTIL_SET_MEMMON
```
```suggestion
$set 60 MSGCAT_UTIL_SET_RKCHECK
```

youngjun9072 replied:
@hornetmj 변경하였습니다.

hornetmj commented on msg/ko_KR.utf8/utils.msg:
```diff
@@ -1322,3 +1322,12 @@ memmon: 데이터베이스 서버의 메모리 사용량 정보를 제공합니
+
+$set 60 MSGCAT_UTIL_SET_MEMMON
+60 \
+rkcheck: HA 복제 키 제약 조건 준수 여부를 점검합니다.\n\
```
NIT:

```suggestion
rkcheck: HA 복제 키 제약 조건 준수 여부를 검사합니다.\n\
```

youngjun9072 replied:
@hornetmj 변경하였습니다.

hornetmj commented on src/executables/utility.h:
```diff
@@ -805,7 +811,8 @@ typedef enum
   FLASHBACK,
   MEMMON,
-  LOGFILEDUMP,
+  CHECK_REPL_CONS,
```
기존에 드린 커멘트가 유틸리티명을 보다 명시적으로 rkcheck에서 check_repl_cons로 변경하자는 의견이였습니다. PM팀 협의 결과 rkcheck로 결정되었기 때문에 관련된 부분 모두 rkcheck로 통일하는게 좋을 것 같습니다.

```suggestion
  RKCHECK,
```

youngjun9072 replied:
@hornetmj 변경하였습니다.

hornetmj commented on src/executables/util_service.c:
```diff
@@ -3205,6 +3211,41 @@ ha_argv_to_args (char *args, int size, const char **argv, HB_PROC_TYPE type)
+	{
+	  continue;
+	}
+
```
HA 모드로 설정된 DB만 rkcheck가 수행되어야 하기 때문에 us_hb_process_copylogdb()와 us_hb_process_applylogdb() 함수를 참고하셔서 다음 코드부 추가 검토해주세요.
```
      prm_set_integer_value (PRM_ID_HA_MODE_FOR_SA_UTILS_ONLY, HA_MODE_FAIL_BACK);
      status = sysprm_load_and_init (dbs[i], NULL, SYSPRM_IGNORE_INTL_PARAMS);
      if (status != NO_ERROR)
        {
          goto ret;
        }

      if (util_get_ha_mode_for_sa_utils () == HA_MODE_OFF)
        {
          continue;
        }
```

youngjun9072 replied:
@hornetmj 추가하였습니다.

hornetmj replied:
이 부분 포함해서 유틸에 대한 검증 시나리오 추가해 주시길 바랍니다.
1. usage 출력
2. 옵션 처리
3. rk, !rk, !rk_fk -> rk_pk, rk_fk -> !rk_pk 등의 조건에서 유틸 동작 검증
4. 에러 처리 검증 - 각 에러 발생 시나리오 및 메시지 출력 검증 등

youngjun9072 replied:
@hornetmj 시나리오및 결과 추가했습니다.

hornetmj commented on src/executables/util_service.c:
```diff
@@ -3205,6 +3211,41 @@ ha_argv_to_args (char *args, int size, const char **argv, HB_PROC_TYPE type)
+	  continue;
+	}
+
+      const char *rkcheck_args[] = { UTIL_ADMIN_NAME, UTIL_RKCHECK,
```
--rk-result와 --fk-result 옵션 전달이 필요해 보입니다.

youngjun9072 replied:
@hornetmj 
recheck_args 배열에 나열한 값을들을 프로세스 실행 시 파라미터로 전달하는걸로 분석하였는데, 
이번에 추가한 get_print_flags () 함수에서 아무 플래그가 설정되어있지 않을 경우 rk와 fk의 결과가 모두 출력되도록 구현하였습니다. 
혹시 터미널과의 동작이 다르다면 말씀부탁드립니다.

hornetmj replied:
다음 동작을 보시면 -r 옵션 사용 시에도 fk 검사가 수행되고 있습니다. -f 옵션도 결과는 같습니다.
```
[hornetmj1] ~
$ cub_admin rkcheck -r demodb

Constraint checks completed for all replication-enabled tables.

$ cat CUBRID/log/demodb_rkcheck_20260120_0434.list
< Replication Key(PRIMARY KEY or NOT NULL UNIQUE KEY) Constraint Violations >
All replicated tables comply with the replication key constraints.


< Foreign Key Constraint Violations >
All replicated tables comply with the foreign key constraints.
```

youngjun9072 replied:
@hornetmj 옵션 입력시 리스트 파일에는 타이틀과 목록이 해당 옵션에대해서만 출력하도록 수정했습니다.

hornetmj commented on src/query/execute_schema.c:
```diff
@@ -8899,6 +8899,35 @@ check_ha_repl_fk_ref_all_replicated (DB_OBJECT * class_obj)
+	  if (do_print)
+	    {
+	      fprintf (fp, "%s -> %s\n", sm_get_ch_name (class_obj),
+		       sm_get_ch_name (ws_mop (&tmp_c->fk_info->ref_class_oid, NULL)));
```
NIT:

ws_mop (&(tmp_c->fk_info->ref_class_oid)를 MOP 로컬 변수로 재사용

youngjun9072 replied:
@hornetmj 변경하였습니다.

hornetmj commented on src/query/execute_schema.c:
```diff
@@ -8899,6 +8899,35 @@ check_ha_repl_fk_ref_all_replicated (DB_OBJECT * class_obj)
+
+  assert (class_obj != NULL);
+
+  for (tmp_c = db_get_constraints (class_obj); tmp_c; tmp_c = db_constraint_next (tmp_c))
```
class_obj 클래스가 REPLICATION=OFF 설정인 경우에 대한 처리가 필요해 보입니다.

youngjun9072 replied:
@hornetmj check_repl_constraint_violations() -> check_fk_constraint() -> check_repl_constraint_violations() -> log_ha_repl_fk_ref_all_replicated() 순서로 호출되고있는데, check_repl_constraint_violations() 함수에서 RK 테이블과 FK테이블에 대해 공통적으로 continue 처리를 해주고있습니다.
```
  for (c = classes; c != NULL; c = c->next)
    {
      if (db_is_system_class (c->op) || !sm_is_replication_class (c->op))
	{
	  continue;
	}

      ret += check_func (c->op, fp, do_print);
    }
```

youngjun9072 committed:
Apply code review feedback

- Remove incorrect return handling in rkcheck()
- Revise log messages
  - Rename result to results to reflect multiple outcomes
  - Add detailed message descriptions
  - Fix catalog set name typo ($set 60 MSGCAT_UTIL_SET_MEMMON -> $set 60 MSGCAT_UTIL_SET_RKCHECK)
- Add missing function declarations
- Add function-level comments
- Rename CHECK_REPL_CONS to RKCHECK
  - Rename HA-invoked utility constant to SC_RKCHECK
- Ensure rkcheck runs only for databases configured in HA mode
- Refactor duplicated function calls by caching results in local variables

hornetmj commented on msg/en_US.utf8/utils.msg:
```diff
@@ -1336,3 +1336,13 @@ usage: %1$s memmon [OPTION] database-name\n\
+\n\
+valid option:\n\
+    -r, --rk-results           print RK(Replication Key constraint violation list)\n\
+    -f, --fk-results           print FK(Foreign Key) constraint violation list\n
```
```suggestion
    -r, --rk-results           print RK(Replication Key) constraint violation list\n\
    -f, --fk-results           print FK(Foreign Key) constraint violation list\n
```

youngjun9072 replied:
@hornetmj 변경했습니다.

hornetmj commented on src/executables/util_cs.c:
```diff
@@ -2727,6 +2777,229 @@ changemode (UTIL_FUNCTION_ARG * arg)
+  print_flags = get_print_flags (arg_map);
+
+  /* error message log file */
+  snprintf (er_msg_file, sizeof (er_msg_file) - 1, "%s_%s.err", database_name, arg->command_name);
```
NIT:
```suggestion
  snprintf (er_msg_file, PATH_MAX, "%s_%s.err", database_name, arg->command_name);
```

youngjun9072 replied:
@hornetmj 변경했습니다,

hornetmj commented on src/executables/util_cs.c:
```diff
@@ -2727,6 +2777,229 @@ changemode (UTIL_FUNCTION_ARG * arg)
+  er_init (er_msg_file, ER_NEVER_EXIT);
+
+  generate_violation_list_file_name (violation_list_file, database_name, arg->command_name);
+  fp = open_violation_list_file (database_name, arg->command_name, violation_list_file, PATH_MAX);
```
generate_violation_list_file_name() 함수가 중복 호출되고 있습니다. 코드 정리가 필요해 보입니다.
1. generate_violation_list_file_name() 호출
2. open_violation_list_file() 함수 내에서 다시 generate_violation_list_file_name() 호출

youngjun9072 replied:
@hornetmj 변경했습니다.

youngjun9072 committed:
Apply code review feedback

- Adjust section title output based on option flags

youngjun9072 committed:
Add comments

hornetmj commented on src/executables/util_cs.c:
```diff
@@ -2727,6 +2800,259 @@ changemode (UTIL_FUNCTION_ARG * arg)
+    }
+
+  rk_violation_count = check_repl_constraint_violations (classes, fp, (print_flags & PRINT_RK) != 0, &rk_check_desc);
+  fk_violation_count = check_repl_constraint_violations (classes, fp, (print_flags & PRINT_FK) != 0, &fk_check_desc);
```
PRINT_RK 또는 PRINT_FK의 경우 단순히 출력 여부를 제어하기 보다는 검사 자체를 제어하는게 맞을 것 같습니다. 현재 -r 옵션을 입력했을 때에도 -f의 검사를 수행하거나 관련 로직 일부가 수행되고 있습니다. 그 반대 역시 마찬가지 입니다.

youngjun9072 replied:
@hornetmj PM 요구사항이 옵션을 지정한 경우 해당 기능만 수행하는것으로 변경됨에따라 코드에 반영하였습니다.

hornetmj commented on src/executables/util_cs.c:
```diff
@@ -2727,6 +2800,259 @@ changemode (UTIL_FUNCTION_ARG * arg)
+
+  /* error message log file */
+  snprintf (er_msg_file, PATH_MAX, "%s_%s.err", database_name, arg->command_name);
+  er_init (er_msg_file, ER_NEVER_EXIT);
```
er_final() 처리가 필요

youngjun9072 replied:
@hornetmj 반영했습니다..

hornetmj commented on src/executables/util_cs.c:
```diff
@@ -2727,6 +2800,259 @@ changemode (UTIL_FUNCTION_ARG * arg)
+  db_set_client_type (DB_CLIENT_TYPE_ADMIN_UTILITY);
+  db_login ("DBA", NULL);
+
+  if (db_restart (arg->command_name, TRUE, database_name))
```
db_shutdown() 처리 필요

youngjun9072 replied:
@hornetmj 반영했습니다.

hornetmj reviewed (commented):
위에 resolve 처리 되지 않았으나 접혀있어서 보이지 않는 부분이 있습니다. 같이 확인 부탁드립니다.

youngjun9072 committed:
Apply code review feedback

- Change option semantics from output control to constraint check control
- Add resource cleanup function

youngjun9072 committed:
Merge remote-tracking branch 'origin/feature/add-replication-option' into ha_start
